File: frontend/bioreact/lib/model/stirrer.dart
import 'package:bench_core/channels.dart';
import 'package:bench_core/log.dart';

class StirrerModel with Logging {
  double omega = 0.0;
  double omegaTarget = 100.0;
  double dutyCycle = 0.0;

  double kInertial = 1e-3;
  double kOmega = 3e3;
  double kProp = 1e-4;

  double tCurrent = 0.0;

  double tOn = 1000.0;
  double tOff = 0.0;
  double tLastSwitch = 0.0;
  bool isOn = true;

  void advanceTime(double tLast, double dt) {
    double tCurrent = tLast + dt;

    if (isOn && (tCurrent - tLastSwitch > tOn) && (tOff > 0)) {
      isOn = false;
      tLastSwitch = tCurrent;
    } else if (!isOn && (tCurrent - tLastSwitch > tOff)) {
      isOn = true;
      tLastSwitch = tCurrent;
    }

    // Compute new state
    if (isOn) {
      dutyCycle += kProp * (omegaTarget - omega) * dt;
      if (dutyCycle < 0.0) {
        dutyCycle = 0.0;
      } else if (dutyCycle > 1.0) {
        dutyCycle = 1.0;
      }
    } else {
      dutyCycle = 0.0;
    }

    omega += 1.0 / kInertial * (dutyCycle - omega / kOmega) * dt;
  }
}

// typedef FloatValueAction = SetValueAction<double>;

class StirrerBus extends ChannelBus {
  MeasurementChannel<double> dutyCycle;
  SetPointChannel speed;
  SetValueChannel<double> onTime;
  SetValueChannel<double> offTime;

  StirrerBus._({
    required this.dutyCycle,
    required this.speed,
    required this.onTime,
    required this.offTime,
  });

  factory StirrerBus() {
    var instance = StirrerBus._(
      dutyCycle: MeasurementChannel(id: "duty_cycle"),
      speed: SetPointChannel(id: "speed"),
      onTime: SetValueChannel(id: "on_time"),
      offTime: SetValueChannel(id: "off_time"),
    );
    return instance;
  }

  @override
  String get busId => "stirrer";

  @override
  List<ChannelBus> get children => [dutyCycle, speed, onTime, offTime];
}

class StirrerDeviceConnector extends DeviceConnector<StirrerBus> {
  StirrerModel device;
  SetPointDeviceConnector speed;
  MeasurementDeviceConnector<double> dutyCycle;
  SetValueDeviceConnector<double> onTime;
  SetValueDeviceConnector<double> offTime;

  StirrerDeviceConnector._(
    this.device, {
    required this.speed,
    required this.dutyCycle,
    required this.onTime,
    required this.offTime,
  });

  factory StirrerDeviceConnector(StirrerModel device) {
    var speedConnector = SetPointDeviceConnector.fn(
      currentValueReader: () => device.omega,
      targetValueReader: () => device.omegaTarget,
      targetValueWriter: (v) => device.omegaTarget = v,
    );

    var dutyCycleConnector =
        MeasurementDeviceConnector<double>(() => device.dutyCycle);

    var onTimeConnector = SetValueDeviceConnector<double>.fn(
      valueReader: () => device.tOn,
      valueWriter: (v) => device.tOn = v,
    );

    var offTimeConnector = SetValueDeviceConnector<double>.fn(
      valueReader: () => device.tOff,
      valueWriter: (v) => device.tOff = v,
    );

    return StirrerDeviceConnector._(device,
        speed: speedConnector,
        dutyCycle: dutyCycleConnector,
        onTime: onTimeConnector,
        offTime: offTimeConnector);
  }

  @override
  void sample() {
    speed.sample();
    dutyCycle.sample();
  }

  @override
  void connectForwardChannels(StirrerBus bus) {
    dutyCycle.connectForwardChannels(bus.dutyCycle);
    speed.connectForwardChannels(bus.speed);
    onTime.connectForwardChannels(bus.onTime);
    offTime.connectForwardChannels(bus.offTime);
  }

  @override
  void connectReverseChannels(StirrerBus bus) {
    speed.connectReverseChannels(bus.speed);
    onTime.connectReverseChannels(bus.onTime);
    offTime.connectReverseChannels(bus.offTime);
  }
}

class StirrerControlConnector extends ControlConnector<StirrerBus> {
  MeasurementControlConnector<double> dutyCycle;
  SetPointControlConnector speed;
  SetValueControlConnector<double> onTime;
  SetValueControlConnector<double> offTime;

  StirrerControlConnector({
    Function(double value)? onDutyCycleValue,
    Function(double value)? onSpeedCurrentValue,
    Function(double value)? onSpeedTargetValue,
    Function(double value)? onOnTimeValue,
    Function(double value)? onOffTimeValue,
  })  : dutyCycle = MeasurementControlConnector<double>(onDutyCycleValue),
        speed = SetPointControlConnector(
            onCurrentValue: onSpeedCurrentValue,
            onTargetReadValue: onSpeedTargetValue),
        onTime = SetValueControlConnector<double>(onOnTimeValue),
        offTime = SetValueControlConnector<double>(onOffTimeValue) {
    speed.currentConnector.configure(
        label: "Stirrer speed", unit: "rpm", formatter: FloatFormatter());
    dutyCycle.configure(label: "Motor load", formatter: PercentFormatter());
  }

  @override
  void connectForwardChannels(StirrerBus bus) {
    speed.connectForwardChannels(bus.speed);
    dutyCycle.connectForwardChannels(bus.dutyCycle);
    onTime.connectForwardChannels(bus.onTime);
    offTime.connectForwardChannels(bus.offTime);
  }

  @override
  void connectReverseChannels(StirrerBus bus) {
    speed.connectReverseChannels(bus.speed);
    onTime.connectReverseChannels(bus.onTime);
    offTime.connectReverseChannels(bus.offTime);
  }

  @override
  void dispose() {
    dutyCycle.dispose();
    speed.dispose();
    onTime.dispose();
    offTime.dispose();
  }
}

File: backend/model/src/model/stirrer.rs
use flowmbed_peripherals::channels::{MeasurementChannel, SetPointChannel, DeviceBusConnector, IOConnector, SetValueChannel};
use flowmbed_peripherals::util::QualifiedPath;
use flowmbed_peripherals::mqtt::MqttService;
use flowmbed_peripherals::set_value_handler;

pub struct StirrerModel {
  omega: f32,
  omega_target: f32,
  duty_cycle: f32,
  
  k_inertial: f32,
  k_omega: f32,
  k_prop: f32,

  on_time: f32,
  off_time: f32,
  t_last_switch: f32,
  is_on: bool,
}

impl StirrerModel {
  pub fn new() -> Self {
    Self { 
      omega: 0.0, 
      omega_target: 100.0, 
      duty_cycle: 0.0, 
      
      k_inertial: 1e-3, 
      k_omega: 3e3, 
      k_prop: 1e-4, 
      
      on_time: 1.0, 
      off_time: 2.0, 
      t_last_switch: 0.0, 
      is_on: true 
    }
  }

  pub fn step(&mut self, t_last: f32, dt: f32) {
    let t_current = t_last + dt;

    if self.is_on && (t_current - self.t_last_switch > self.on_time * 60.0) && (self.off_time > 0.0) {
      self.is_on = false;
      self.t_last_switch = t_current;
    } else if !self.is_on && (t_current - self.t_last_switch > self.off_time * 60.0) {
      self.is_on = true;
      self.t_last_switch = t_current;
    }

    // Compute new state
    if self.is_on {
      self.duty_cycle += self.k_prop * (self.omega_target - self.omega) * dt;
      if self.duty_cycle < 0.0 {
        self.duty_cycle = 0.0;
      } else if self.duty_cycle > 1.0 {
        self.duty_cycle = 1.0;
      }
    } else {
      self.duty_cycle = 0.0;
    }

    self.omega += 1.0 / self.k_inertial * (self.duty_cycle - self.omega / self.k_omega) * dt;

  }
}

pub struct StirrerBus {
  id: String,
  speed: SetPointChannel,
  duty_cycle: MeasurementChannel<f32>,
  on_time: SetValueChannel<f32> ,
  off_time: SetValueChannel<f32> ,
}

impl StirrerBus {
  pub fn new(id: &str) -> Self {
    Self { 
      id: id.to_owned(), 
      speed: SetPointChannel::new("speed"), 
      duty_cycle: MeasurementChannel::new("duty_cycle"), 
      on_time: SetValueChannel::new("on_time"), 
      off_time: SetValueChannel::new("off_time") 
    }
  }
}

impl DeviceBusConnector<StirrerModel> for StirrerBus {
  fn sample(&self, device: &StirrerModel) {
    self.speed.current.sample(device.omega);
    self.duty_cycle.sample(device.duty_cycle);
  }
  fn handle_actions(&self, device: &mut StirrerModel) {
    set_value_handler!(self, on_time, device, on_time);
    set_value_handler!(self, off_time, device, off_time);
    set_value_handler!(self, speed.target, device, omega_target);
  }
}

impl IOConnector for StirrerBus {  
  fn connect_io(&mut self, comm: &mut dyn MqttService, qpath: &QualifiedPath) {
      self.speed.connect_io(comm, &qpath.append(&self.id));
      self.duty_cycle.connect_io(comm, &qpath.append(&self.id));
      self.on_time.connect_io(comm, &qpath.append(&self.id));
      self.off_time.connect_io(comm, &qpath.append(&self.id));     
  }
}